pkg/cli/server/server.go

```diff
@@ -311,3 +332,32 @@ func getArgValueFromList(searchArg string, argList []string) string {
 	}
 	return value
 }
+
+func getServerURLFromEtcd(ctx context.Context, serverConfig *server.Config, agentConfig *cmds.Agent) {
+	// setting LBServerPort to a prespecified port to initialize the kubeconfigs with the right address
+	agentConfig.LBServerPort = lbServerPort
+	// start a thread to check for the server ip if set from etcd
+	if serverConfig.ControlConfig.HTTPSPort != serverConfig.ControlConfig.SupervisorPort {
+		go setServerURLTmp(ctx, serverConfig, agentConfig)
+		return
+	}
+	setServerURLTmp(ctx, serverConfig, agentConfig)
+	agentConfig.ServerURL = agentConfig.ServerURLTmp
+}
```

Comment:
@galal-hussein can you walk us through what's going on here? This func seems to be mis-named as there is no value being returned and then there is the conditional future side-effect that has me worried.

Reply:
sorry it maybe misnamed a little bit, so this function is called by the server's embedded agent when k3s server starts, it has two modes:

1- for rke2 (when httpsPort != supervisorPort) then it runs in a goroutine and once it gets rke2/etcd/serveraddresses it sets the serverURLTmp

2- for k3s and it runs in a blocking loop until it gets the ip from etcd k3s/etcd/serveraddresses and then it sets the serverURLTmp 

note that for rke2 mode we dont block to allow kubelet to start and start etcd since every component is a static pod

Reply:
We have other goroutines reading the serverUrlTmp without a lock?

Reply:
I wonder if this functionality might be better served by establishing a `sync.Once` private var that calls `once.Do` with a  func that blocks

Reply:
I am +1 on refactoring this to something with a SyncOnce, and use a closure (anonymous function) instead of the somewhat awkwardly named `setServerURLTmp` function?

Reply:
fixed